File: backend/app/schemas/pet.py
from pydantic import BaseModel, Field
from typing import Optional, List, Dict
from datetime import datetime


class PetCreate(BaseModel):
    """领养宠物请求"""
    name: str = Field(default="小伙伴", min_length=1, max_length=50)
    species: str = Field(default="pikachu", pattern="^(pikachu|eevee|bulbasaur|charmander|squirtle|jigglypuff|gastly|dratini|machop|abra|geodude|vulpix|growlithe|magikarp|oddish|poliwag)$")


class PetResponse(BaseModel):
    """宠物信息响应"""
    id: int
    user_id: int
    name: str
    species: str
    level: int
    experience: int
    happiness: int
    hunger: int
    evolution_stage: int
    xp_to_next_level: int
    evolution_stage_name: str
    food_balance: int = 0
    last_fed_at: Optional[datetime] = None
    last_interaction_at: Optional[datetime] = None
    created_at: datetime

    class Config:
        from_attributes = True


class PetFeedResponse(BaseModel):
    """喂食响应"""
    message: str
    pet: PetResponse
    leveled_up: bool = False
    evolved: bool = False
    new_level: Optional[int] = None
    new_stage: Optional[int] = None


class PetEventResponse(BaseModel):
    """宠物事件响应"""
    id: int
    pet_id: int
    event_type: str
    detail: Optional[str] = None
    created_at: datetime

    class Config:
        from_attributes = True


class PetLeaderboardEntry(BaseModel):
    """排行榜条目"""
    rank: int
    username: str
    pet_name: str
    species: str
    level: int
    evolution_stage: int
    evolution_stage_name: str


class PetLeaderboardResponse(BaseModel):
    """排行榜响应"""
    entries: List[PetLeaderboardEntry]
    my_rank: Optional[int] = None


class EarnFoodRequest(BaseModel):
    """做题赚粮请求"""
    score: int = Field(ge=0, description="答对题数")
    total: int = Field(ge=1, description="总题数")
    mode: str = Field(pattern="^(flashcard|quiz|fillblank|spelling)$", description="练习模式")


class EarnFoodResponse(BaseModel):
    """赚粮响应"""
    food_earned: int
    food_balance: int
    is_first_today: bool
    breakdown: Dict[str, int]

File: frontend/src/api/pet.ts
import api from './client';

// ========================================
// 宠物养成系统 API
// ========================================

export interface Pet {
  id: number;
  user_id: number;
  name: string;
  species: string;
  level: number;
  experience: number;
  happiness: number;
  hunger: number;
  evolution_stage: number;
  xp_to_next_level: number;
  xp_per_feed: number;
  evolution_stage_name: string;
  food_balance: number;
  current_hp: number;
  is_injured: boolean;
  is_active: boolean;
  last_fed_at: string | null;
  last_interaction_at: string | null;
  created_at: string;
}

export interface PetFeedResponse {
  message: string;
  pet: Pet;
  leveled_up: boolean;
  evolved: boolean;
  new_level: number | null;
  new_stage: number | null;
}

export interface PetEvent {
  id: number;
  pet_id: number;
  event_type: string;
  detail: string | null;
  created_at: string;
}

export interface CreatePetRequest {
  name: string;
  species: string;
}

export interface SwitchPetRequest {
  pet_id: number;
}

export interface PetCollection {
  pets: Pet[];
  active_pet_id: number | null;
  learned_words: number;
  unlocked_slots: number;
  used_slots: number;
  max_slots: number;
  words_per_slot: number;
  next_slot_words: number | null;
  recovery_goal_words: number | null;
  recovery_words_remaining: number;
}

export interface EarnFoodRequest {
  score: number;
  total: number;
  mode: 'classify' | 'quiz' | 'fillblank' | 'spelling';
}

export interface EarnFoodResponse {
  food_earned: number;
  food_balance: number;
  is_first_today: boolean;
  breakdown: {
    base: number;
    accuracy_bonus: number;
    mode_bonus: number;
    daily_bonus: number;
  };
}

export const getMyPet = async (): Promise<Pet | null> => {
  try {
    return await api.get('/student/pet');
  } catch (err: unknown) {
    const status = (err as { response?: { status?: number } })?.response?.status;
    if (status === 404) return null;
    throw err;
  }
};

export const createPet = async (data: CreatePetRequest): Promise<Pet> => {
  return api.post('/student/pet', data);
};

export const getPetCollection = async (): Promise<PetCollection> => {
  return api.get('/student/pet/collection');
};

export const switchPet = async (data: SwitchPetRequest): Promise<Pet> => {
  return api.post('/student/pet/switch', data);
};

export const feedPet = async (): Promise<PetFeedResponse> => {
  return api.post('/student/pet/feed');
};

export const earnFood = async (data: EarnFoodRequest): Promise<EarnFoodResponse> => {
  return api.post('/student/pet/earn-food', data);
};

export const getPetEvents = async (): Promise<PetEvent[]> => {
  return api.get('/student/pet/events');
};

// ========== 排行榜 ==========

export interface PetLeaderboardEntry {
  rank: number;
  username: string;
  pet_name: string;
  species: string;
  level: number;
  evolution_stage: number;
  evolution_stage_name: string;
}

export interface PetLeaderboardResponse {
  entries: PetLeaderboardEntry[];
  my_rank: number | null;
}

export const getPetLeaderboard = async (): Promise<PetLeaderboardResponse> => {
  return api.get('/student/pet/leaderboard');
};
